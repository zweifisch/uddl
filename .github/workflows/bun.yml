name: Bun

on:
  push:
    branches: [ "main" ]
  pull_request:
    branches: [ "main" ]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v4
    - name: Setup Bun
      uses: oven-sh/setup-bun@v2.0.1
    - name: Install Dependencies
      run: bun install --frozen-lockfile
    - name: Run tests
      run: bun test
