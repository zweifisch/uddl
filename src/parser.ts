import { createToken, Lexer, CstParser  } from "chevrotain"

const LCurly = createToken({ name: "LCurly", pattern: /{/ })
const RCurly = createToken({ name: "RCurly", pattern: /}/ })
const LBrace = createToken({ name: "LBrace", pattern: /\(/ })
const RBrace = createToken({ name: "RBrace", pattern: /\)/ })
const Optional = createToken({ name: "Optional", pattern: /\?/ })
const Colon = createToken({ name: "Colon", pattern: /:/ })
const Comma = createToken({ name: "Comma", pattern: /,/ })
const Identifier = createToken({ name: "Identifier", pattern: /[a-zA-Z_][a-zA-Z0-9_]*/ })
const Number = createToken({ name: "Number", pattern: /\d+/ })
const LineBreak = createToken({ name: "LineBreak", pattern: /[\r\n]+/, group: Lexer.SKIPPED })
const WhiteSpace = createToken({ name: "WhiteSpace", pattern: /\s+/, group: Lexer.SKIPPED });

const allTokens = [ LineBreak, WhiteSpace, LCurly, RCurly, LBrace, RBrace, Optional, Colon, Comma, Identifier, Number ]


class SchemaParser extends CstParser {
  constructor() {
    super(allTokens)

    this.RULE("file", () => {
      this.MANY(() => this.SUBRULE((this as any).entry))
    })

    this.RULE("entry", () => {
      this.CONSUME(Identifier)
      this.CONSUME(LCurly)
      this.OPTION(() => this.MANY(() => this.SUBRULE((this as any).property)))
      this.CONSUME(RCurly)
    })

    this.RULE("property", () => {
      this.CONSUME(Identifier)
      this.OPTION(() => {
        this.CONSUME(LBrace)
        this.OPTION2(() => this.MANY(() => this.SUBRULE((this as any).attribute)))
        this.CONSUME(RBrace)
      })
      this.OPTION3(() => this.CONSUME(Optional))
      this.CONSUME(Colon)
      this.CONSUME2(Identifier)
      this.OPTION4(() => this.CONSUME(Comma))
    })

    this.RULE("attribute", () => {
      this.CONSUME(Identifier)
      this.OPTION(() => {
        this.CONSUME(Colon)
        this.SUBRULE((this as any).value)
      })
    })

    this.RULE("value", () => {
      this.OR([
        { ALT: () => this.CONSUME(Number) },
        { ALT: () => this.CONSUME(Identifier) },
      ])
    })

    this.performSelfAnalysis()
  }
}

const lexer = new Lexer(allTokens)
const parser = new SchemaParser()

interface Token {
  image: string,
  startColumn: number,
  startLine: number,
  startOffset: number,
  endColumn: number,
  endLine: number,
  endOffset: number,
  tokenType: {
    name: string
  },
}

interface Node<T> {
  name: string
  children: T
}

export type File = Node<{entry: Array<EntryNode>}>
type EntryNode = Node<{Identifier: [Token], property: Array<PropertyNode>}>
type PropertyNode = Node<{
  Identifier: [Token, Token],
  attribute?: Array<AttributeNode>,
  Optional?: Array<Token>}
>
type AttributeNode = Node<{Identifier: [Token], value?: [ValueNode]}>
type ValueNode = Node<{Number: [Token]} | {Identifier: [Token]}>


export function parse(input: string): File {
  const {errors, tokens} = lexer.tokenize(input)

  if (errors.length > 0) {
    throw errors
  }

  parser.input = tokens
  const result = (parser as any).file()

  if (parser.errors.length > 0) {
    throw parser.errors[0]
  }

  return result
}

export function formatError(input: string, error: any) {
  if (typeof Array.isArray(error)) {
    return error.map((x: { message: string, line: number, column: number }) =>
      (x.line > 2 ? input.split("\n")[x.line - 2] + "\n" : '') +
      input.split('\n')[x.line - 1] + "\n"
      + " ".repeat(x.column - 1) + "^ "
      + x.message
    ).join("\n")
  }
  if (error.token) {
    const { message, token: { startLine, startColumn } } = error as { message: string, token: { startLine: number, startColumn: number } }
    return [
      startLine > 2 && input.split("\n")[startLine - 2],
      input.split("\n")[startLine - 1],
      " ".repeat(startColumn - 1) + "^ ",
      `${message} at ${startLine},${startColumn}`,
    ].filter(x => x).join('\n')
  }
  return error
}

export function isNumber(value: ValueNode): value is Node<{Number: [Token]}> {
  return 'Number' in value.children
} 

export function isIdentifier(value: ValueNode): value is Node<{Identifier: [Token]}> {
  return 'Identifier' in value.children
} 
