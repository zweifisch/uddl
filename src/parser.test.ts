import {test, expect} from 'bun:test'
import { parse } from './parser'

function skipRexExp(input: any): any {
  if (Array.isArray(input)) {
    return input.map(skipRexExp)
  }
  if (input instanceof RegExp) {
    return undefined
  }
  if (input !== null && typeof input === 'object') {
    return Object.fromEntries(Object.entries(input).map(([k, v]) => [k, skipRexExp(v)]))
  }
  return input
}

test('parse', () => {

  const input = `
User {
  id(primary_key auto_increment): int,
  name?: text,
  email(unique): text,
  gender(default:0): int,
}
`
  expect(skipRexExp(parse(input))).toMatchSnapshot()
})