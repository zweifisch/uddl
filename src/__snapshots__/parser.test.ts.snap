// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parse 1`] = `
{
  "children": {
    "entry": [
      {
        "children": {
          "Identifier": [
            {
              "endColumn": 4,
              "endLine": 2,
              "endOffset": 4,
              "image": "User",
              "startColumn": 1,
              "startLine": 2,
              "startOffset": 1,
              "tokenType": {
                "CATEGORIES": [],
                "PATTERN": undefined,
                "categoryMatches": [],
                "categoryMatchesMap": {},
                "isParent": false,
                "name": "Identifier",
                "tokenTypeIdx": 10,
              },
              "tokenTypeIdx": 10,
            },
          ],
          "LCurly": [
            {
              "endColumn": 6,
              "endLine": 2,
              "endOffset": 6,
              "image": "{",
              "startColumn": 6,
              "startLine": 2,
              "startOffset": 6,
              "tokenType": {
                "CATEGORIES": [],
                "PATTERN": undefined,
                "categoryMatches": [],
                "categoryMatchesMap": {},
                "isParent": false,
                "name": "LCurly",
                "tokenTypeIdx": 3,
              },
              "tokenTypeIdx": 3,
            },
          ],
          "RCurly": [
            {
              "endColumn": 1,
              "endLine": 7,
              "endOffset": 111,
              "image": "}",
              "startColumn": 1,
              "startLine": 7,
              "startOffset": 111,
              "tokenType": {
                "CATEGORIES": [],
                "PATTERN": undefined,
                "categoryMatches": [],
                "categoryMatchesMap": {},
                "isParent": false,
                "name": "RCurly",
                "tokenTypeIdx": 4,
              },
              "tokenTypeIdx": 4,
            },
          ],
          "property": [
            {
              "children": {
                "Colon": [
                  {
                    "endColumn": 33,
                    "endLine": 3,
                    "endOffset": 40,
                    "image": ":",
                    "startColumn": 33,
                    "startLine": 3,
                    "startOffset": 40,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Colon",
                      "tokenTypeIdx": 8,
                    },
                    "tokenTypeIdx": 8,
                  },
                ],
                "Comma": [
                  {
                    "endColumn": 38,
                    "endLine": 3,
                    "endOffset": 45,
                    "image": ",",
                    "startColumn": 38,
                    "startLine": 3,
                    "startOffset": 45,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Comma",
                      "tokenTypeIdx": 9,
                    },
                    "tokenTypeIdx": 9,
                  },
                ],
                "Identifier": [
                  {
                    "endColumn": 4,
                    "endLine": 3,
                    "endOffset": 11,
                    "image": "id",
                    "startColumn": 3,
                    "startLine": 3,
                    "startOffset": 10,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Identifier",
                      "tokenTypeIdx": 10,
                    },
                    "tokenTypeIdx": 10,
                  },
                  {
                    "endColumn": 37,
                    "endLine": 3,
                    "endOffset": 44,
                    "image": "int",
                    "startColumn": 35,
                    "startLine": 3,
                    "startOffset": 42,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Identifier",
                      "tokenTypeIdx": 10,
                    },
                    "tokenTypeIdx": 10,
                  },
                ],
                "LBrace": [
                  {
                    "endColumn": 5,
                    "endLine": 3,
                    "endOffset": 12,
                    "image": "(",
                    "startColumn": 5,
                    "startLine": 3,
                    "startOffset": 12,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "LBrace",
                      "tokenTypeIdx": 5,
                    },
                    "tokenTypeIdx": 5,
                  },
                ],
                "RBrace": [
                  {
                    "endColumn": 32,
                    "endLine": 3,
                    "endOffset": 39,
                    "image": ")",
                    "startColumn": 32,
                    "startLine": 3,
                    "startOffset": 39,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "RBrace",
                      "tokenTypeIdx": 6,
                    },
                    "tokenTypeIdx": 6,
                  },
                ],
                "attribute": [
                  {
                    "children": {
                      "Identifier": [
                        {
                          "endColumn": 16,
                          "endLine": 3,
                          "endOffset": 23,
                          "image": "primary_key",
                          "startColumn": 6,
                          "startLine": 3,
                          "startOffset": 13,
                          "tokenType": {
                            "CATEGORIES": [],
                            "PATTERN": undefined,
                            "categoryMatches": [],
                            "categoryMatchesMap": {},
                            "isParent": false,
                            "name": "Identifier",
                            "tokenTypeIdx": 10,
                          },
                          "tokenTypeIdx": 10,
                        },
                      ],
                    },
                    "name": "attribute",
                  },
                  {
                    "children": {
                      "Identifier": [
                        {
                          "endColumn": 31,
                          "endLine": 3,
                          "endOffset": 38,
                          "image": "auto_increment",
                          "startColumn": 18,
                          "startLine": 3,
                          "startOffset": 25,
                          "tokenType": {
                            "CATEGORIES": [],
                            "PATTERN": undefined,
                            "categoryMatches": [],
                            "categoryMatchesMap": {},
                            "isParent": false,
                            "name": "Identifier",
                            "tokenTypeIdx": 10,
                          },
                          "tokenTypeIdx": 10,
                        },
                      ],
                    },
                    "name": "attribute",
                  },
                ],
              },
              "name": "property",
            },
            {
              "children": {
                "Colon": [
                  {
                    "endColumn": 8,
                    "endLine": 4,
                    "endOffset": 54,
                    "image": ":",
                    "startColumn": 8,
                    "startLine": 4,
                    "startOffset": 54,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Colon",
                      "tokenTypeIdx": 8,
                    },
                    "tokenTypeIdx": 8,
                  },
                ],
                "Comma": [
                  {
                    "endColumn": 14,
                    "endLine": 4,
                    "endOffset": 60,
                    "image": ",",
                    "startColumn": 14,
                    "startLine": 4,
                    "startOffset": 60,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Comma",
                      "tokenTypeIdx": 9,
                    },
                    "tokenTypeIdx": 9,
                  },
                ],
                "Identifier": [
                  {
                    "endColumn": 6,
                    "endLine": 4,
                    "endOffset": 52,
                    "image": "name",
                    "startColumn": 3,
                    "startLine": 4,
                    "startOffset": 49,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Identifier",
                      "tokenTypeIdx": 10,
                    },
                    "tokenTypeIdx": 10,
                  },
                  {
                    "endColumn": 13,
                    "endLine": 4,
                    "endOffset": 59,
                    "image": "text",
                    "startColumn": 10,
                    "startLine": 4,
                    "startOffset": 56,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Identifier",
                      "tokenTypeIdx": 10,
                    },
                    "tokenTypeIdx": 10,
                  },
                ],
                "Optional": [
                  {
                    "endColumn": 7,
                    "endLine": 4,
                    "endOffset": 53,
                    "image": "?",
                    "startColumn": 7,
                    "startLine": 4,
                    "startOffset": 53,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Optional",
                      "tokenTypeIdx": 7,
                    },
                    "tokenTypeIdx": 7,
                  },
                ],
              },
              "name": "property",
            },
            {
              "children": {
                "Colon": [
                  {
                    "endColumn": 16,
                    "endLine": 5,
                    "endOffset": 77,
                    "image": ":",
                    "startColumn": 16,
                    "startLine": 5,
                    "startOffset": 77,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Colon",
                      "tokenTypeIdx": 8,
                    },
                    "tokenTypeIdx": 8,
                  },
                ],
                "Comma": [
                  {
                    "endColumn": 22,
                    "endLine": 5,
                    "endOffset": 83,
                    "image": ",",
                    "startColumn": 22,
                    "startLine": 5,
                    "startOffset": 83,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Comma",
                      "tokenTypeIdx": 9,
                    },
                    "tokenTypeIdx": 9,
                  },
                ],
                "Identifier": [
                  {
                    "endColumn": 7,
                    "endLine": 5,
                    "endOffset": 68,
                    "image": "email",
                    "startColumn": 3,
                    "startLine": 5,
                    "startOffset": 64,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Identifier",
                      "tokenTypeIdx": 10,
                    },
                    "tokenTypeIdx": 10,
                  },
                  {
                    "endColumn": 21,
                    "endLine": 5,
                    "endOffset": 82,
                    "image": "text",
                    "startColumn": 18,
                    "startLine": 5,
                    "startOffset": 79,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Identifier",
                      "tokenTypeIdx": 10,
                    },
                    "tokenTypeIdx": 10,
                  },
                ],
                "LBrace": [
                  {
                    "endColumn": 8,
                    "endLine": 5,
                    "endOffset": 69,
                    "image": "(",
                    "startColumn": 8,
                    "startLine": 5,
                    "startOffset": 69,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "LBrace",
                      "tokenTypeIdx": 5,
                    },
                    "tokenTypeIdx": 5,
                  },
                ],
                "RBrace": [
                  {
                    "endColumn": 15,
                    "endLine": 5,
                    "endOffset": 76,
                    "image": ")",
                    "startColumn": 15,
                    "startLine": 5,
                    "startOffset": 76,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "RBrace",
                      "tokenTypeIdx": 6,
                    },
                    "tokenTypeIdx": 6,
                  },
                ],
                "attribute": [
                  {
                    "children": {
                      "Identifier": [
                        {
                          "endColumn": 14,
                          "endLine": 5,
                          "endOffset": 75,
                          "image": "unique",
                          "startColumn": 9,
                          "startLine": 5,
                          "startOffset": 70,
                          "tokenType": {
                            "CATEGORIES": [],
                            "PATTERN": undefined,
                            "categoryMatches": [],
                            "categoryMatchesMap": {},
                            "isParent": false,
                            "name": "Identifier",
                            "tokenTypeIdx": 10,
                          },
                          "tokenTypeIdx": 10,
                        },
                      ],
                    },
                    "name": "attribute",
                  },
                ],
              },
              "name": "property",
            },
            {
              "children": {
                "Colon": [
                  {
                    "endColumn": 20,
                    "endLine": 6,
                    "endOffset": 104,
                    "image": ":",
                    "startColumn": 20,
                    "startLine": 6,
                    "startOffset": 104,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Colon",
                      "tokenTypeIdx": 8,
                    },
                    "tokenTypeIdx": 8,
                  },
                ],
                "Comma": [
                  {
                    "endColumn": 25,
                    "endLine": 6,
                    "endOffset": 109,
                    "image": ",",
                    "startColumn": 25,
                    "startLine": 6,
                    "startOffset": 109,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Comma",
                      "tokenTypeIdx": 9,
                    },
                    "tokenTypeIdx": 9,
                  },
                ],
                "Identifier": [
                  {
                    "endColumn": 8,
                    "endLine": 6,
                    "endOffset": 92,
                    "image": "gender",
                    "startColumn": 3,
                    "startLine": 6,
                    "startOffset": 87,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Identifier",
                      "tokenTypeIdx": 10,
                    },
                    "tokenTypeIdx": 10,
                  },
                  {
                    "endColumn": 24,
                    "endLine": 6,
                    "endOffset": 108,
                    "image": "int",
                    "startColumn": 22,
                    "startLine": 6,
                    "startOffset": 106,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "Identifier",
                      "tokenTypeIdx": 10,
                    },
                    "tokenTypeIdx": 10,
                  },
                ],
                "LBrace": [
                  {
                    "endColumn": 9,
                    "endLine": 6,
                    "endOffset": 93,
                    "image": "(",
                    "startColumn": 9,
                    "startLine": 6,
                    "startOffset": 93,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "LBrace",
                      "tokenTypeIdx": 5,
                    },
                    "tokenTypeIdx": 5,
                  },
                ],
                "RBrace": [
                  {
                    "endColumn": 19,
                    "endLine": 6,
                    "endOffset": 103,
                    "image": ")",
                    "startColumn": 19,
                    "startLine": 6,
                    "startOffset": 103,
                    "tokenType": {
                      "CATEGORIES": [],
                      "PATTERN": undefined,
                      "categoryMatches": [],
                      "categoryMatchesMap": {},
                      "isParent": false,
                      "name": "RBrace",
                      "tokenTypeIdx": 6,
                    },
                    "tokenTypeIdx": 6,
                  },
                ],
                "attribute": [
                  {
                    "children": {
                      "Colon": [
                        {
                          "endColumn": 17,
                          "endLine": 6,
                          "endOffset": 101,
                          "image": ":",
                          "startColumn": 17,
                          "startLine": 6,
                          "startOffset": 101,
                          "tokenType": {
                            "CATEGORIES": [],
                            "PATTERN": undefined,
                            "categoryMatches": [],
                            "categoryMatchesMap": {},
                            "isParent": false,
                            "name": "Colon",
                            "tokenTypeIdx": 8,
                          },
                          "tokenTypeIdx": 8,
                        },
                      ],
                      "Identifier": [
                        {
                          "endColumn": 16,
                          "endLine": 6,
                          "endOffset": 100,
                          "image": "default",
                          "startColumn": 10,
                          "startLine": 6,
                          "startOffset": 94,
                          "tokenType": {
                            "CATEGORIES": [],
                            "PATTERN": undefined,
                            "categoryMatches": [],
                            "categoryMatchesMap": {},
                            "isParent": false,
                            "name": "Identifier",
                            "tokenTypeIdx": 10,
                          },
                          "tokenTypeIdx": 10,
                        },
                      ],
                      "value": [
                        {
                          "children": {
                            "Number": [
                              {
                                "endColumn": 18,
                                "endLine": 6,
                                "endOffset": 102,
                                "image": "0",
                                "startColumn": 18,
                                "startLine": 6,
                                "startOffset": 102,
                                "tokenType": {
                                  "CATEGORIES": [],
                                  "PATTERN": undefined,
                                  "categoryMatches": [],
                                  "categoryMatchesMap": {},
                                  "isParent": false,
                                  "name": "Number",
                                  "tokenTypeIdx": 11,
                                },
                                "tokenTypeIdx": 11,
                              },
                            ],
                          },
                          "name": "value",
                        },
                      ],
                    },
                    "name": "attribute",
                  },
                ],
              },
              "name": "property",
            },
          ],
        },
        "name": "entry",
      },
    ],
  },
  "name": "file",
}
`;
